<a href="https://colab.research.google.com/github/hrishikeshchavan99/Black-Friday-Dataset/blob/master/EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Exploratory Data Analysis for Black Friday Dataset**

In [0]:
import pandas as pd

In [0]:
df = pd.read_csv("/content/drive/My Drive/Colab Notebooks/train.csv")

In [3]:
df.head()

User_ID Product_ID Gender  ... Product_Category_2  Product_Category_3 Purchase
0  1000001  P00069042      F  ...                NaN                 NaN     8370
1  1000001  P00248942      F  ...                6.0                14.0    15200
2  1000001  P00087842      F  ...                NaN                 NaN     1422
3  1000001  P00085442      F  ...               14.0                 NaN     1057
4  1000002  P00285442      M  ...                NaN                 NaN     7969

[5 rows x 12 columns]

This can be also implemented by 

In [4]:
df.loc[0:4]

User_ID Product_ID Gender  ... Product_Category_2  Product_Category_3 Purchase
0  1000001  P00069042      F  ...                NaN                 NaN     8370
1  1000001  P00248942      F  ...                6.0                14.0    15200
2  1000001  P00087842      F  ...                NaN                 NaN     1422
3  1000001  P00085442      F  ...               14.0                 NaN     1057
4  1000002  P00285442      M  ...                NaN                 NaN     7969

[5 rows x 12 columns]

In [5]:
df.tail()

User_ID Product_ID  ... Product_Category_3 Purchase
550063  1006033  P00372445  ...                NaN      368
550064  1006035  P00375436  ...                NaN      371
550065  1006036  P00375436  ...                NaN      137
550066  1006038  P00375436  ...                NaN      365
550067  1006039  P00371644  ...                NaN      490

[5 rows x 12 columns]

In [6]:
df.shape

(550068, 12)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     550068 non-null  int64  
 1   Product_ID                  550068 non-null  object 
 2   Gender                      550068 non-null  object 
 3   Age                         550068 non-null  object 
 4   Occupation                  550068 non-null  int64  
 5   City_Category               550068 non-null  object 
 6   Stay_In_Current_City_Years  550068 non-null  object 
 7   Marital_Status              550068 non-null  int64  
 8   Product_Category_1          550068 non-null  int64  
 9   Product_Category_2          376430 non-null  float64
 10  Product_Category_3          166821 non-null  float64
 11  Purchase                    550068 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 50.4+ MB


In [8]:
df.describe()

User_ID     Occupation  ...  Product_Category_3       Purchase
count  5.500680e+05  550068.000000  ...       166821.000000  550068.000000
mean   1.003029e+06       8.076707  ...           12.668243    9263.968713
std    1.727592e+03       6.522660  ...            4.125338    5023.065394
min    1.000001e+06       0.000000  ...            3.000000      12.000000
25%    1.001516e+06       2.000000  ...            9.000000    5823.000000
50%    1.003077e+06       7.000000  ...           14.000000    8047.000000
75%    1.004478e+06      14.000000  ...           16.000000   12054.000000
max    1.006040e+06      20.000000  ...           18.000000   23961.000000

[8 rows x 7 columns]

#Removing null/missing values

to check the total number of missing values in each column

In [9]:
df.isnull().sum()

User_ID                            0
Product_ID                         0
Gender                             0
Age                                0
Occupation                         0
City_Category                      0
Stay_In_Current_City_Years         0
Marital_Status                     0
Product_Category_1                 0
Product_Category_2            173638
Product_Category_3            383247
Purchase                           0
dtype: int64

There are total 173638 null values in Product_Category_2 and 383247 null values in Product_Category_3            
To handle missing numerical values

Fill missing values backward or forward<br>
Forward fill:

In [0]:
df = df.fillna(method = 'pad')

Check if all null values removed or not

In [11]:
df.isnull().sum()

User_ID                       0
Product_ID                    0
Gender                        0
Age                           0
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
Product_Category_1            0
Product_Category_2            1
Product_Category_3            1
Purchase                      0
dtype: int64

In [12]:
df[['Product_Category_2', 'Product_Category_3']].head()


Product_Category_2  Product_Category_3
0                 NaN                 NaN
1                 6.0                14.0
2                 6.0                14.0
3                14.0                14.0
4                14.0                14.0

Still one null value left each in product category 2 and 3<br>
So remove it using backward fill

In [0]:
df = df.fillna(method = 'bfill')

In [14]:
df.isnull().sum()

User_ID                       0
Product_ID                    0
Gender                        0
Age                           0
Occupation                    0
City_Category                 0
Stay_In_Current_City_Years    0
Marital_Status                0
Product_Category_1            0
Product_Category_2            0
Product_Category_3            0
Purchase                      0
dtype: int64

So all null values have been removed

Check again with assert that there are no missing values in the dataframe to gain confidence



In [0]:
assert pd.notnull(df).all().all()

As it is not throwing any assertion error, all null values have been removed

#General Operations<br>
indexing and slicing 

In [16]:
# select first row of dataframe
df.loc[0]

User_ID                         1000001
Product_ID                    P00069042
Gender                                F
Age                                0-17
Occupation                           10
City_Category                         A
Stay_In_Current_City_Years            2
Marital_Status                        0
Product_Category_1                    3
Product_Category_2                    6
Product_Category_3                   14
Purchase                           8370
Name: 0, dtype: object

In [17]:
#selecting last row of dataframe¶
df.iloc[-1]

User_ID                         1006039
Product_ID                    P00371644
Gender                                F
Age                               46-50
Occupation                            0
City_Category                         B
Stay_In_Current_City_Years           4+
Marital_Status                        1
Product_Category_1                   20
Product_Category_2                    2
Product_Category_3                   11
Purchase                            490
Name: 550067, dtype: object

In [18]:
#selecting first column of dataframe¶
df.iloc[:,0]

0         1000001
1         1000001
2         1000001
3         1000001
4         1000002
           ...   
550063    1006033
550064    1006035
550065    1006036
550066    1006038
550067    1006039
Name: User_ID, Length: 550068, dtype: int64

In [19]:
#selecting last column of dataframe¶
df.iloc[:,-1]

0          8370
1         15200
2          1422
3          1057
4          7969
          ...  
550063      368
550064      371
550065      137
550066      365
550067      490
Name: Purchase, Length: 550068, dtype: int64

In [20]:
#selecting first ten rows for a specific column
df.loc[:,'Product_ID'].head(10)

0    P00069042
1    P00248942
2    P00087842
3    P00085442
4    P00285442
5    P00193542
6    P00184942
7    P00346142
8     P0097242
9    P00274942
Name: Product_ID, dtype: object

In [21]:
#Selecting random five rows for multiple columns
df.loc[[10, 21, 52, 73, 94],['Age','Occupation']]

Age  Occupation
10  26-35          20
21  26-35          12
52  46-50           1
73  18-25           3
94   0-17          10

In [22]:
#selecting random rows with random columns
df.iloc[[10,24,49], [0,5,8]]

User_ID City_Category  Product_Category_1
10  1000005             A                   5
24  1000008             C                   1
49  1000011             C                   8

In [23]:
#selecting random rows and columns in range
df.iloc[0:6, 5:8]

City_Category Stay_In_Current_City_Years  Marital_Status
0             A                          2               0
1             A                          2               0
2             A                          2               0
3             A                          2               0
4             C                         4+               0
5             A                          3               0

Getting index of max and min values

In [24]:
df['Purchase'].idxmax()

87440

In [25]:
df['Purchase'].idxmin()

545946

In [26]:
# getting the row with the minimum Purchase value 

df.loc[df['Purchase'].idxmax()]

User_ID                         1001474
Product_ID                    P00052842
Gender                                M
Age                               26-35
Occupation                            4
City_Category                         A
Stay_In_Current_City_Years            2
Marital_Status                        1
Product_Category_1                   10
Product_Category_2                   15
Product_Category_3                    8
Purchase                          23961
Name: 87440, dtype: object

Indexing a single value with at() and iat()

In [27]:
# get value at 114321st row and Purchase column pair

df.at[114321, 'Purchase']

5367

In [28]:
# get value at 1st row and 11th column pair

df.iat[1, 11]

15200

Boolean operations

In [29]:
# get the purchase amount with a given user_id and product_id

df.loc[((df['User_ID'] == 1000001) & (df['Product_ID'] == 'P00069042')), 'Purchase']

0    8370
Name: Purchase, dtype: int64

checking values with isin method

In [30]:
df.head(6)

User_ID Product_ID Gender  ... Product_Category_2  Product_Category_3 Purchase
0  1000001  P00069042      F  ...                6.0                14.0     8370
1  1000001  P00248942      F  ...                6.0                14.0    15200
2  1000001  P00087842      F  ...                6.0                14.0     1422
3  1000001  P00085442      F  ...               14.0                14.0     1057
4  1000002  P00285442      M  ...               14.0                14.0     7969
5  1000003  P00193542      M  ...                2.0                14.0    15227

[6 rows x 12 columns]

In [31]:
values=[1000001,'P00085442','F',0-17,10,'A',2,0,3,6,14,1422]

df_indexed=df.isin(values)

df_indexed.head(6)

User_ID  Product_ID  ...  Product_Category_3  Purchase
0     True       False  ...                True     False
1     True       False  ...                True     False
2     True       False  ...                True      True
3     True        True  ...                True     False
4    False       False  ...                True     False
5    False       False  ...                True     False

[6 rows x 12 columns]